src/operations/execute_operation.ts

```diff
@@ -66,45 +62,48 @@ export interface ExecutionResult {
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(topologyProvider: TopologyProvider, operation: T): Promise<TResult>;
+>(client: MongoClient, operation: T): Promise<TResult>;
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(topologyProvider: TopologyProvider, operation: T, callback: Callback<TResult>): void;
+>(client: MongoClient, operation: T, callback: Callback<TResult>): void;
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(
-  topologyProvider: TopologyProvider,
-  operation: T,
-  callback?: Callback<TResult>
-): Promise<TResult> | void;
+>(client: MongoClient, operation: T, callback?: Callback<TResult>): Promise<TResult> | void;
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(
-  topologyProvider: TopologyProvider,
-  operation: T,
-  callback?: Callback<TResult>
-): Promise<TResult> | void {
+>(client: MongoClient, operation: T, callback?: Callback<TResult>): Promise<TResult> | void {
   if (!(operation instanceof AbstractOperation)) {
     // TODO(NODE-3483): Extend MongoRuntimeError
     throw new MongoRuntimeError('This method requires a valid operation instance');
   }
 
   return maybePromise(callback, callback => {
-    let topology: Topology;
-    try {
-      // TODO(NODE-4151): Use skipPingOnConnect and call connect here to make client.connect optional
-      topology = getTopology(topologyProvider);
-    } catch (error) {
-      return callback(error);
+    const topology = client.topology;
+
+    if (topology == null) {
+      if (client.s.hasBeenClosed) {
+        return callback(
+          new MongoNotConnectedError('Client must be connected before running operations')
+        );
+      }
+      client.s.options[Symbol.for('@@mdb.skipPingOnConnect')] = true;
```

Comment:
Just a question - is this feature flag still needed for those who still explicitly calling `client.connect()`?

Reply:
No one needs to set this, it was just for us to use in tests to make them consistent for other drivers. I'm then using it here to make sure the auto connect doesn't run two operations.